pkg/workers/notifications/notification.go

```diff
@@ -0,0 +1,211 @@
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/cozy/cozy-stack/pkg/config"
+	"github.com/cozy/cozy-stack/pkg/jobs"
+
+	multierror "github.com/hashicorp/go-multierror"
+
+	fcm "github.com/appleboy/go-fcm"
+	apns "github.com/sideshow/apns2"
+	apns_cert "github.com/sideshow/apns2/certificate"
+	"github.com/sideshow/apns2/payload"
+	apns_token "github.com/sideshow/apns2/token"
+)
+
+const (
+	// Android platform using Firebase Cloud Messaging (FCM)
+	Android = "android"
+	// IOS platform using APNS/2
+	IOS = "ios"
+)
+
+var (
+	fcmClient *fcm.Client
+	iosClient *apns.Client
+)
+
+func init() {
+	jobs.AddWorker("notification", &jobs.WorkerConfig{
+		Concurrency:  runtime.NumCPU(),
+		MaxExecCount: 2,
+		Timeout:      10 * time.Second,
+		WorkerInit:   Init,
+		WorkerFunc:   Worker,
+	})
+}
+
+// Message contains a push notification request.
+type Message struct {
+	Platform    string `json:"platform"`
+	DeviceToken string `json:"device_token"`
+	Topic       string `json:"topic,omitempty"`
+	Title       string `json:"title,omitempty"`
+	Message     string `json:"message,omitempty"`
+	Priority    string `json:"priority,omitempty"`
+	Sound       string `json:"sound,omitempty"`
+
+	Data map[string]interface{} `json:"data,omitempty"`
+}
+
+// Init initializes the necessary global clients
+func Init() (err error) {
+	conf := config.GetConfig().Notifications
+
+	if conf.AndroidAPIKey != "" {
+		fcmClient, err = fcm.NewClient(config.GetConfig().Notifications.AndroidAPIKey)
+		if err != nil {
+			return
+		}
+	}
+
+	if conf.IOSCertificateKeyPath != "" {
+		var authKey *ecdsa.PrivateKey
+		var certificateKey tls.Certificate
+
+		switch filepath.Ext(conf.IOSCertificateKeyPath) {
+		case ".p12":
+			certificateKey, err = apns_cert.FromP12File(conf.IOSCertificateKeyPath, conf.IOSCertificatePassword)
+		case ".pem":
+			certificateKey, err = apns_cert.FromPemFile(conf.IOSCertificateKeyPath, conf.IOSCertificatePassword)
+		case ".p8":
+			authKey, err = apns_token.AuthKeyFromFile(conf.IOSCertificateKeyPath)
+		default:
+			err = errors.New("wrong certificate key extension")
+		}
+		if err != nil {
+			return err
+		}
+
+		if authKey != nil {
+			t := &apns_token.Token{
+				AuthKey: authKey,
+				KeyID:   conf.IOSKeyID,
+				TeamID:  conf.IOSTeamID,
+			}
+			iosClient = apns.NewTokenClient(t)
+		} else {
+			iosClient = apns.NewClient(certificateKey).Development()
+		}
+		if conf.Development {
+			iosClient = iosClient.Development()
+		} else {
+			iosClient = iosClient.Production()
+		}
+	}
+	return
+}
+
+// Worker is the worker that just logs its message (useful for debugging)
+func Worker(ctx *jobs.WorkerContext) error {
+	var msg Message
+	if err := ctx.UnmarshalMessage(&msg); err != nil {
+		return err
+	}
+	switch msg.Platform {
+	case Android:
+		return pushToAndroid(ctx, &msg)
+	case IOS:
+		return pushToIOS(ctx, &msg)
+	default:
+		return fmt.Errorf("notifications: unknown platform %q", msg.Platform)
```

Comment:
`%q` -> is it a typo?

Reply:
No, it formats as a quoted string: `fmt.Sprintf("my name is %q", "pierre") -> "my name is \"pierre\""`